pennylane/io/qualtran_io.py

```diff
@@ -0,0 +1,208 @@
+                    reg.name: np.empty((*reg.shape, reg.bitsize), dtype=object).flatten()
+                    for reg in binst.bloq.signature.rights()
+                }
+
+                # This is to track the number of wires defined at the LeftDangle stage
+                # so if we need to add more wires, we know what index to start at
+                soq_to_wires_len = 0
+                if len(soq_to_wires.values()) > 0:
+                    try:
+                        soq_to_wires_len = list(soq_to_wires.values())[-1][-1] + 1
+                    except (IndexError, TypeError):
+                        soq_to_wires_len = list(soq_to_wires.values())[-1] + 1
+
+                for pred in pred_cxns:
+                    soq = pred.right
+                    soq_to_wires[soq] = soq_to_wires[pred.left]
+                    in_quregs[soq.reg.name][soq.idx] = soq_to_wires[soq]
+
+                for succ in succ_cxns:
+                    soq = succ.left
+                    if soq.reg.side == Side.RIGHT:
+                        # When in_quregs != out_quregs, it means that there are wires unaccounted
+                        # for. We account for these wires and update soq_to_wires and in_quregs
+                        # accordingly.
+                        if len(in_quregs) != len(out_quregs):
+                            total_elements = np.prod(soq.reg.shape) * soq.reg.bitsize
+                            ascending_vals = np.arange(
+                                soq_to_wires_len,
+                                total_elements + soq_to_wires_len,
+                                dtype=object,
+                            )
+                            in_quregs[soq.reg.name] = ascending_vals.reshape(
+                                (*soq.reg.shape, soq.reg.bitsize)
+                            )
+                        soq_to_wires[soq] = in_quregs[soq.reg.name][soq.idx]
+
+                total_wires = [w for ws in in_quregs.values() for w in list(ws.flatten())]
+
+                mapped_wires = [wires[idx] for idx in total_wires]
+                op = binst.bloq.as_pl_op(mapped_wires)
+
+                if op:
+                    ops.append(op)
+        except (DecomposeNotImplementedError, DecomposeTypeError):
+            pass
+
+        return ops
+
+    # pylint: disable=invalid-overridden-method, arguments-renamed
+    @property
+    def has_matrix(self) -> bool:  # pylint: disable=protected-access
+        r"""Return if the bloq has a valid matrix representation."""
+        bloq = self._hyperparameters["bloq"]
+        matrix = bloq.tensor_contract()
+        return matrix.shape == (2 ** len(self.wires), 2 ** len(self.wires))
+
+    def compute_matrix(
+        *params, **kwargs
+    ):  # pylint: disable=unused-argument, no-self-argument, no-method-argument, protected-access
+        bloq = params[0]._hyperparameters["bloq"]
+        matrix = bloq.tensor_contract()
+        return matrix
```

Comment:
For bloqs that have `Allocate()`, this will actually return the matrix acting on |00....0>; maybe raise error?